lib/galaxy/web/framework/__init__.py

```diff
@@ -4,4 +4,14 @@
 
 from . import base
 
-url_for = base.routes.url_for
+
+def handle_url_for(*args, **kargs) -> str:
+    try:
+        return base.routes.url_for(*args, **kargs)
+    except AttributeError:
+        if len(args) > 0:
+            return args[0]
```

Comment:
I am not 100% sure if this is the best solution or just a dirty hack :)

But apparently, this doctest:
https://github.com/galaxyproject/galaxy/blob/57d66b9b226ccfb4d7769ee48c139cffbc731b2c/lib/galaxy/tools/parameters/basic.py#L766

was always assuming that an exception will raise here:
https://github.com/galaxyproject/galaxy/blob/57d66b9b226ccfb4d7769ee48c139cffbc731b2c/lib/galaxy/tools/parameters/basic.py#L782

and, in that situation, the test was expecting the first parameter of `url_for` to be returned.

This should fix the failing test and since we are deprecating this `url_for` when moving to FastAPI the actual value may not be important, but I will sleep better if someone with more knowledge about the implications of this `<param type='baseurl'>` can verify if not returning a proper URL here is ok.

Reply:
```
tools/data_source/ucsc_tablebrowser.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner" />

tools/data_source/mousemine.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=mousemine" />

tools/data_source/eupathdb.xml
11:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=eupathdb" />

tools/data_source/biomart.xml
20:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner/biomart" />

tools/data_source/ucsc_tablebrowser_test.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner" />

tools/data_source/intermine.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=intermine" />

tools/data_source/cbi_rice_mart.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner/biomart" />

tools/data_source/worm_modencode.xml
12:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=modENCODEworm" />

tools/data_source/wormbase.xml
12:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=wormbase" />

tools/data_source/ucsc_tablebrowser_archaea.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner" />

tools/data_source/fly_modencode.xml
12:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=modENCODEfly" />

tools/data_source/ratmine.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=ratmine" />

tools/data_source/gramene_mart.xml
20:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner/biomart" />

tools/data_source/biomart_test.xml
20:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner/biomart" />

tools/data_source/wormbase_test.xml
12:		<param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=wormbase_test" />

tools/data_source/flymine.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=flymine" />

tools/data_source/hapmapmart.xml
24:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner/hapmapmart" />
 
tools/data_source/flymine_test.xml
17:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=flymine" />

tools/data_source/sra.xml
25:        <param name="GALAXY_URL" type="baseurl" value="/tool_runner?tool_id=sra_source" />
```

It does look like it is used. I think we need a FastAPI method of capturing at least the base URL for requests in some context global and using it when dealing with these methods. It looks like we really only need the base to prefix not all of the logic of routes.

Reply:
Thank you! I will investigate a bit how to get this base URL globally from FastAPI and report back here.

Reply:
Awesome - thanks so much!

Reply:
It seems the way to get the `url_for` in a FastAPI context is either by using `app.url_path_for(...)`, `router.url_path_for(...)` or `request.url_for(...)` where `app: fastapi.FastApp`, `router: fastapi.APIRouter` and `request: fastapi.Request`.

After spending some time looking into this I couldn't come up with a solution to access the `app` or the current `request` instance globally without the feeling that I am doing something really evil... so I am stuck here again 😅 

Any pythonic ideas for this?

Reply:
I'm thinking about something like this https://github.com/galaxyproject/galaxy/pull/11904 - I'm not sure if it is great or works though :sweat_smile: .

Reply:
That looks way better! :)